### PR TITLE
GeneralImprovements

### PR DESCRIPTION
**Pull request by ZharkovIlia**
In general, there is performance improvement. Some bugs are fixed and some redundancy is eliminated.

### CODE REVIEW

File: src/mongo/client/dbclient_rs.cpp [hunk 1]
```diff
@@ -124,7 +124,7 @@ ReadPreferenceSetting* _extractReadPref(const BSONObj& query, int queryOptions)
     }
 
     // Default read pref is primary only or secondary preferred with slaveOK
-    ReadPreference pref = queryOptions & QueryOption_SlaveOk
+    ReadPreference pref = (queryOptions & QueryOption_SlaveOk)
```
Comment (acmorrow) [inline src/mongo/client/dbclient_rs.cpp]: Bitwise and has a higher precedence than the ternery operator, so I don't think this code needed to change.

File: src/mongo/client/connection_pool.cpp [hunk 2]
```diff
@@ -110,7 +110,7 @@ void ConnectionPool::closeAllInUseConnections() {
 void ConnectionPool::_cleanUpStaleHosts_inlock(Date_t now) {
     if (now > _lastCleanUpTime + kCleanUpInterval) {
         for (HostLastUsedMap::iterator itr = _lastUsedHosts.begin(); itr != _lastUsedHosts.end();
-             itr++) {
+             ++itr) {
```
Comment (acmorrow) [inline src/mongo/client/connection_pool.cpp]: When the result is not captured/used, the definition of the increment operator is inline, it seems unlikely that the compiler will generate different code for ++itr and itr++. Can you actually demonstrate a performance or codegen difference from this or any of the other identical changes?

File: src/mongo/client/dbclient_rs_test.cpp [hunk 3]
```diff
@@ -98,7 +98,7 @@ class BasicRS : public unittest::Test {
 
 void assertOneOfNodesSelected(MockReplicaSet* replSet,
                               ReadPreference rp,
-                              const std::vector<std::string> hostNames) {
+                              const std::vector<std::string>& hostNames) {
```
Comment (acmorrow) [inline src/mongo/client/dbclient_rs_test.cpp]: While you aren't wrong that this is wasteful, this is also in a test. I don't see that it matters.

File: src/mongo/db/catalog/collection.cpp [hunk 4]
```diff
@@ -1129,7 +1129,7 @@ class RecordStoreValidateAdaptor : public ValidateAdaptor {
 
                 if (indexEntryCount != 0) {
                     indexEntryCount--;
-                    dassert(indexEntryCount >= 0);
+                    //dassert(indexEntryCount >= 0);
```
Comment (acmorrow) [inline src/mongo/db/catalog/collection.cpp]: - Why did you comment out the dassert
- If the dassert should be removed, please just delete it, rather than commenting it out.

File: src/mongo/client/native_sasl_client_session.cpp [hunk 5]
```diff
@@ -38,7 +38,7 @@
 namespace mongo {
 namespace {
 
-SaslClientSession* createNativeSaslClientSession(const std::string mech) {
+SaslClientSession* createNativeSaslClientSession(const std::string& mech) {
```
Comment (acmorrow) [inline src/mongo/client/native_sasl_client_session.cpp]: @spencerjackson - Why is this parameter unused?

File: src/mongo/db/auth/auth_index_d.cpp [hunk 6]
```diff
@@ -100,7 +100,7 @@ Status verifySystemIndexes(OperationContext* txn) {
     std::vector<IndexDescriptor*> indexes;
     indexCatalog->findIndexesByKeyPattern(txn, v1SystemUsersKeyPattern, false, &indexes);
 
-    if (indexCatalog && !indexes.empty()) {
+    if (!indexes.empty()) {
```
Comment (acmorrow) [inline src/mongo/db/auth/auth_index_d.cpp]: I presume you are removing this because if indexCatalog was null, line 99 above would be undefined behavior, and therefore we don't need the check?

File: src/mongo/db/commands/feature_compatibility_version.cpp [hunk 7]
```diff
@@ -207,8 +207,6 @@ void FeatureCompatibilityVersion::set(OperationContext* txn, StringData version)
         // featureCompatibilityVersion to 3.4. This is a new index version that isn't supported by
         // versions of MongoDB earlier than 3.4 that will cause 3.2 secondaries to crash when it is
         // replicated.
-        std::vector<BSONObj> indexSpecs{k32IncompatibleIndexSpec};
```
Comment (acmorrow) [inline src/mongo/db/commands/feature_compatibility_version.cpp]: @tessavitabile - Did this declaration (and the identical on below) become unused as some point? I presume that is why it has been deleted here.

File: src/mongo/db/concurrency/lock_manager.cpp [hunk 8]
```diff
@@ -229,7 +229,7 @@ struct LockHead {
 
     // Methods to maintain the granted queue
     void incGrantedModeCount(LockMode mode) {
-        invariant(grantedCounts[mode] >= 0);
+        //invariant(grantedCounts[mode] >= 0);
```
Comment (acmorrow) [inline src/mongo/db/concurrency/lock_manager.cpp]: Why did you remove this invariant? Do you believe it will never fire? If you can prove that, just delete it, don't comment it out.

File: src/mongo/db/concurrency/lock_manager.cpp [hunk 9]
```diff
@@ -246,7 +246,7 @@ struct LockHead {
 
     // Methods to maintain the conflict queue
     void incConflictModeCount(LockMode mode) {
-        invariant(conflictCounts[mode] >= 0);
+        //invariant(conflictCounts[mode] >= 0);
```
Comment (acmorrow) [inline src/mongo/db/concurrency/lock_manager.cpp]: Same here re invariant.

File: src/mongo/db/pipeline/granularity_rounder_preferred_numbers_test.cpp [hunk 10]
```diff
@@ -51,7 +51,7 @@ const vector<string> preferredNumberSeries{
  * we check to see if the values are approximately equal with tolerance 'delta'.
  */
 void testEquals(Value actual, Value expected, double delta = DELTA) {
-    if (actual.getType() == BSONType::NumberDouble || actual.getType() == BSONType::NumberDouble) {
+    if (actual.getType() == BSONType::NumberDouble || expected.getType() == BSONType::NumberDouble) {
```
Comment (acmorrow) [inline src/mongo/db/pipeline/granularity_rounder_preferred_numbers_test.cpp]: Nice catch. How did you find this?

File: src/mongo/db/range_deleter.cpp [hunk 11]
```diff
@@ -129,7 +129,7 @@ static void logCursorsWaiting(RangeDeleteEntry* entry) {
 }
 
 struct RangeDeleter::NSMinMax {
-    NSMinMax(std::string ns, const BSONObj min, const BSONObj max) : ns(ns), min(min), max(max) {}
+    NSMinMax(const std::string& ns, const BSONObj min, const BSONObj max) : ns(ns), min(min), max(max) {}
```
Comment (acmorrow) [inline src/mongo/db/range_deleter.cpp]: A better change here might be to leave the parameter as by-value, and just re-write the member initializer as ```ns(std::move(ns))```

File: src/mongo/db/repl/data_replicator_test.cpp [hunk 12]
```diff
@@ -131,7 +131,7 @@ class DataReplicatorTest : public executor::ThreadPoolExecutorTest, public SyncS
         return _syncSourceSelector->shouldChangeSyncSource(currentSource, metadata);
     }
 
-    void scheduleNetworkResponse(std::string cmdName, const BSONObj& obj) {
+    void scheduleNetworkResponse(const std::string& cmdName, const BSONObj& obj) {
```
Comment (acmorrow) [inline src/mongo/db/repl/data_replicator_test.cpp]: I don't care too much about eliminating a string copy inside a unit test.

File: src/mongo/db/repl/rs_rollback.cpp [hunk 13]
```diff
@@ -125,7 +125,7 @@ namespace {
 
 class RSFatalException : public std::exception {
 public:
-    RSFatalException(std::string m = "replica set fatal exception") : msg(m) {}
+    RSFatalException(const std::string& m = "replica set fatal exception") : msg(m) {}
```
Comment (acmorrow) [inline src/mongo/db/repl/rs_rollback.cpp]: Much like above, you could just leave the parameter as by-value, and then write the initailizer for msg as ```msg(std::move(m))```.

File: src/mongo/db/storage/mmap_v1/mmap_windows.cpp [hunk 14]
```diff
@@ -185,7 +185,7 @@ void* MemoryMappedFile::map(const char* filenameIn, unsigned long long& length)
     filename[255] = 0;
     {
         size_t len = strlen(filename);
-        for (size_t i = len - 1; i >= 0; i--) {
+        for (int i = (len > 0 ? static_cast<int>(len - 1) : -1); i >= 0; i--) {
```
Comment (acmorrow) [inline src/mongo/db/storage/mmap_v1/mmap_windows.cpp]: Maybe simpler to just only execute the for loop if len > 0?

File: src/mongo/dbtests/jstests.cpp [hunk 15]
```diff
@@ -1439,7 +1439,7 @@ class DeeplyNestedObject : public TestRoundTrip {
         } else {
             std::stringstream ss;
             ss << "{\"" << depth << "\":" << buildJson(depth - 1) << "}";
-            depth--;
+            depth--; // For developers: Is this operation redundant? Is something missed?
```
Comment (acmorrow) [inline src/mongo/dbtests/jstests.cpp]: Why do you think it is redundant?

File: src/mongo/util/net/sockaddr.cpp [hunk 16]
```diff
@@ -144,7 +144,7 @@ bool SockAddr::isLocalHost() const {
         default:
             return false;
     }
-    fassert(16502, false);
+    //fassert(16502, false);
```
Comment (acmorrow) [inline src/mongo/util/net/sockaddr.cpp]: If you want to remove code, remove it. Even better here would be to write MONGO_UNREACHABLE, in which case you can probably remove the ```return false``` as well, since MONGO_UNREACHABLE has ```noreturn``` semantics.

**Comment by ramonfm** [comment]
@ZharkovIlia , please note that you'll need to sign the [contributor agreement](https://www.mongodb.com/legal/contributor-agreement) before we can consider incorporating any of your changes into our tree.

File: src/mongo/client/native_sasl_client_session.cpp (hunk 5 above)
Comment (spencerjackson) [inline src/mongo/client/native_sasl_client_session.cpp]: Polymorphism. SaslClientSession::create is set to point to this function. Another implementation, createCyrusSaslClientSession, uses the parameter.

File: src/mongo/client/dbclient_rs.cpp (hunk 1 above)
Comment (ZharkovIlia) [inline src/mongo/client/dbclient_rs.cpp]: I only assumed that it can improve code readability. Of course, ignore this change, if you want.

File: src/mongo/client/connection_pool.cpp (hunk 2 above)
Comment (ZharkovIlia) [inline src/mongo/client/connection_pool.cpp]: I can demonstrate a codegen difference only when compiling via gcc without optimizations.

File: src/mongo/client/dbclient_rs_test.cpp (hunk 3 above)
Comment (ZharkovIlia) [inline src/mongo/client/dbclient_rs_test.cpp]: I've understood.

File: src/mongo/db/auth/auth_index_d.cpp (hunk 6 above)
Comment (ZharkovIlia) [inline src/mongo/db/auth/auth_index_d.cpp]: 'indexCatalog' is assigned a pointer to a field of object 'collection' of class 'Collection', so it cannot be null.

File: src/mongo/db/commands/feature_compatibility_version.cpp (hunk 7 above)
Comment (ZharkovIlia) [inline src/mongo/db/commands/feature_compatibility_version.cpp]: indexSpecs variable is never used.

File: src/mongo/db/pipeline/granularity_rounder_preferred_numbers_test.cpp (hunk 10 above)
Comment (ZharkovIlia) [inline src/mongo/db/pipeline/granularity_rounder_preferred_numbers_test.cpp]: I just study one of static analyzer and I chose your project at random.

File: src/mongo/db/catalog/collection.cpp (hunk 4 above)
Comment (ZharkovIlia) [inline src/mongo/db/catalog/collection.cpp]: indexEntryCount is unsigned integer variable. If it can be negative in your mind consider using signed integer variable.

File: src/mongo/db/concurrency/lock_manager.cpp (hunk 8 above)
Comment (ZharkovIlia) [inline src/mongo/db/concurrency/lock_manager.cpp]: grantedCounts is an array of unsigned integers.

File: src/mongo/db/concurrency/lock_manager.cpp (hunk 9 above)
Comment (ZharkovIlia) [inline src/mongo/db/concurrency/lock_manager.cpp]: Same as in the example above.

File: src/mongo/db/range_deleter.cpp (hunk 11 above)
Comment (ZharkovIlia) [inline src/mongo/db/range_deleter.cpp]: Yes, I agree with you.

File: src/mongo/db/storage/mmap_v1/mmap_windows.cpp (hunk 14 above)
Comment (ZharkovIlia) [inline src/mongo/db/storage/mmap_v1/mmap_windows.cpp]: I am only concerned about comparison between unsigned integer variable and 0. Even simpler it may be if len will have type int, because size of 'filename' is not more than 256.

File: src/mongo/dbtests/jstests.cpp (hunk 15 above)
Comment (ZharkovIlia) [inline src/mongo/dbtests/jstests.cpp]: 'depth' is a function parameter and is never used after line 1441.

File: src/mongo/util/net/sockaddr.cpp (hunk 16 above)
Comment (ZharkovIlia) [inline src/mongo/util/net/sockaddr.cpp]: I will take it into account.

**Comment by ZharkovIlia** [comment]
@acmorrow , Is there anything that prevents you from acceptance this pull request?

File: src/mongo/client/connection_pool.cpp (hunk 2 above)
Comment (acmorrow) [inline src/mongo/client/connection_pool.cpp]: OK. Then I would prefer that you back out the iterator changes, everywhere.

File: src/mongo/client/dbclient_rs.cpp (hunk 1 above)
Comment (acmorrow) [inline src/mongo/client/dbclient_rs.cpp]: Please back this change out.

File: src/mongo/client/dbclient_rs_test.cpp (hunk 3 above)
Comment (acmorrow) [inline src/mongo/client/dbclient_rs_test.cpp]: Please unwind the change then

File: src/mongo/db/auth/auth_index_d.cpp (hunk 6 above)
Comment (acmorrow) [inline src/mongo/db/auth/auth_index_d.cpp]: Yes, we are saying the same thing then. This change is fine.

File: src/mongo/db/commands/feature_compatibility_version.cpp (hunk 7 above)
Comment (acmorrow) [inline src/mongo/db/commands/feature_compatibility_version.cpp]: ping @tessavitabile re the unused indexSpecs variable. Also, I would have expected a compiler diagnostic if these are actually unused.

**Comment by acmorrow** [comment]
@ZharkovIlia - Have you signed the contributor agreement?

File: src/mongo/db/commands/feature_compatibility_version.cpp (hunk 7 above)
Comment (acmorrow) [inline src/mongo/db/commands/feature_compatibility_version.cpp]: No, you can leave this one for now.

**Comment by ZharkovIlia** [comment]
@acmorrow - Yes, I have signed the contributor agreement.

**Comment by ramonfm** [comment]
Thanks for signing the contributor agreement @ZharkovIlia 
 
I see you've submitted #1122 , does that PR replaced this one? If so, can we close this one to avoid confusion?

Thanks,
Ramón.

**Comment by ZharkovIlia** [comment]
Yes, of course.

**Comment by ramonfm** [comment]
Thanks, closing now.
